### PR TITLE
Proposal API's were added

### PR DESCRIPTION
**Pull request by ahmedrns**
Get proposal list API is added
Get proposal by id API is added
Get proposal pre evaluation is added

### CODE REVIEW

File: src/Stoa.ts [hunk 1]
```diff
@@ -264,6 +266,8 @@ class Stoa extends WebService {
         this.app.get("/holder/:address", isBlackList, this.getBoaHolder.bind(this));
         this.app.get("/average_fee_chart", isBlackList, this.averageFeeChart.bind(this));
         this.app.get("/search/hash/:hash", isBlackList, this.searchHash.bind(this));
+        this.app.get("/proposals/", isBlackList, this.getProposals.bind(this));
+        this.app.get("/proposalbyid/:proposal_id", isBlackList, this.getProposalById.bind(this));
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
        this.app.get("/proposal/:proposal_id", isBlackList, this.getProposalById.bind(this));
```

File: src/Stoa.ts [hunk 2]
```diff
@@ -2471,6 +2475,89 @@ class Stoa extends WebService {
...
+                });
+                return res.status(500).send("Failed to data lookup");
+            });
+    }
+
+    /* Get proposal by id
+     * @returns Returns proposals of the ledger.
+     */
+    public async getProposalById(req: express.Request, res: express.Response) {
+        const proposal_id = String(req.params.proposal_id);
```
Comment (TrustHenry) [inline src/Stoa.ts]: Proposal_id must be checked.

File: src/Types.ts [hunk 3]
```diff
@@ -888,6 +888,141 @@ export interface IEmitBlock {
...
+     */
+    proposal_tx_hash: string;
+
+    /**
+     * Proposer name 
+     */
+    proposer_name: string;
+
+    /**
+     * Proposal fee tx hash
```
Comment (TrustHenry) [inline src/Types.ts]:
```suggestion
     * Proposal Fee transaction hash
```

File: src/Types.ts [hunk 4]
```diff
@@ -888,6 +888,141 @@ export interface IEmitBlock {
...
+     */
+    proposal_fee_tx_hash: string;
+
+    /**
+     * Proposal fee
+     */
+    proposal_fee: number;
+
+    /**
+     * Proposal start date
```
Comment (TrustHenry) [inline src/Types.ts]:
```suggestion
     * Proposal voting start date
```

File: src/Types.ts [hunk 5]
```diff
@@ -888,6 +888,141 @@ export interface IEmitBlock {
...
+     */
+    voting_start_date: number;
+
+    /**
+     * Proposal end date
```
Comment (TrustHenry) [inline src/Types.ts]:
```suggestion
     * Proposal voting end date
```

File: src/Stoa.ts [hunk 6]
```diff
@@ -2471,6 +2475,89 @@ class Stoa extends WebService {
...
+                        proposer_name: data[0].proposer_name,
+                        fund_amount: data[0].fund_amount,
+                        proposal_fee: data[0].proposal_fee,
+                        proposal_type: data[0].proposal_type,
+                        vote_start_height: data[0].vote_start_height,
+                        voting_start_date: data[0].voting_start_date,
+                        vote_end_height: data[0].vote_end_height,
+                        voting_end_date: data[0].voting_end_date,
+                        proposal_status: data[0].proposal_status,
+                        proposal_date: data[0].submit_time
```
Comment (TrustHenry) [inline src/Stoa.ts]: I think there are data item that are missing.
Please check the screen again.

**Review by TrustHenry — CHANGES_REQUESTED** [review]

**Comment by TrustHenry** [comment]
This branch has conflicts that must be resolved

File: src/Stoa.ts [hunk 7]
```diff
@@ -2481,6 +2531,93 @@ class Stoa extends WebService {
...
+            return;
+        }
+        this.ledger_storage
+            .getProposalById(proposal_id)
+            .then((data: any[]) => {
+                if (data.length === 0) {
+                    return res.status(204).send(`The data does not exist.`);
+                } else {
+                    let proposal: IProposalAPI =
+                    {
```
Comment (TrustHenry) [inline src/Stoa.ts:2544]: app_name,
vote_fee,
proposal_fee_tx_hash,
proposer_address,
proposal_fee_address,
submit_time,
ave_pre_evaluation_score,
pre_evaluation_start_time,
pre_evaluation_end_time,
assess_node_count,
assess_average_score,
assess_completeness_score,
assess_realization_score,
assess_profitability_score,
assess_attractiveness_score,
assess_expansion_score

It needs to add this item.

File: src/modules/storage/LedgerStorage.ts [hunk 8]
```diff
@@ -3532,6 +3573,65 @@ export class LedgerStorage extends Storages {
...
+                    P.proposal_type,
+                    P.fund_amount,
+                    P.vote_start_height,
+                    P.vote_end_height,
+                    P.proposal_status,
+                    M.submit_time,
+                    M.proposer_name,
+                    count(*) OVER() AS full_count
+                    FROM proposal P
+                    INNER JOIN proposal_metadata M
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: Why is use INNER JOIN?
Proposal_metadata data may not exist.

File: src/modules/storage/LedgerStorage.ts [hunk 9]
```diff
@@ -3532,6 +3573,65 @@ export class LedgerStorage extends Storages {
...
+                    P.proposal_fee,
+                    P.proposal_type,
+                    P.vote_start_height,
+                    M.voting_start_date,
+                    P.vote_end_height,
+                    M.voting_end_date,
+                    M.submit_time,
+                    P.proposal_status
+                FROM proposal P 
+                INNER JOIN proposal_metadata M
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: Why is use INNER JOIN?
Proposal_metadata data may not exist.

File: src/modules/storage/LedgerStorage.ts (hunk 8 above)
Comment (ahmedrns) [reply src/modules/storage/LedgerStorage.ts]: We need to display data from proposal_metadata table, the data is for frontend UI.

File: src/modules/storage/LedgerStorage.ts (hunk 9 above)
Comment (ahmedrns) [reply src/modules/storage/LedgerStorage.ts]: We need to display data from proposal_metadata table, the data is for frontend UI.

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: Then use outer join.

File: src/Stoa.ts (hunk 7 above)
Comment (ahmedrns) [reply src/Stoa.ts:2544]: Pre-evaluation data is in separate API, according to the UI design we don't need addition fields

**Review by TrustHenry — APPROVED** [review]

**Pull request merged**